File: lib/model/protobuf_message_codec.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:protobuf/protobuf.dart';

import 'protos/messaging.pb.dart';

class ProtobufMessageCodec extends StandardMessageCodec {
  static const int _valueTimestamp = 30;
  static const int _valueContact = 31;
  static const int _valueMessage = 32;
  static const int _valueConversation = 33;

  void writeValue(WriteBuffer buffer, dynamic value) {
    if (value == null) {
      super.writeValue(buffer, value);
      return;
    }

    int type = 0;
    switch (value.runtimeType) {
      case Timestamp:
        type = _valueTimestamp;
        break;
      case Contact:
        type = _valueContact;
        break;
      case Message:
        type = _valueMessage;
        break;
      case Conversation:
        type = _valueConversation;
        break;
    }

    if (type == 0) {
      super.writeValue(buffer, value);
      return;
    }

    buffer.putUint8(type);
    var serialized = (value as GeneratedMessage).writeToBuffer();
    writeSize(buffer, serialized.length);
    buffer.putUint8List(serialized);
  }

  dynamic readValueOfType(int type, ReadBuffer buffer) {
    if (type < 30) {
      return super.readValueOfType(type, buffer);
    }

    final int length = readSize(buffer);
    var serialized = buffer.getUint8List(length);
    switch (type) {
      case _valueTimestamp:
        return Conversation.fromBuffer(serialized);
      case _valueContact:
        return Contact.fromBuffer(serialized);
      case _valueMessage:
        return Message.fromBuffer(serialized);
      case _valueConversation:
        return Conversation.fromBuffer(serialized);
      default:
        throw FormatException("Unknown data type $type");
    }
  }
}

File: android/app/src/main/kotlin/io/lantern/isimud/model/ProtobufMessageCodec.kt
package io.lantern.isimud.model

import android.nfc.FormatException
import com.google.protobuf.GeneratedMessageLite
import io.flutter.plugin.common.StandardMessageCodec
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer


class ProtobufMessageCodec: StandardMessageCodec() {
    override fun writeValue(stream: ByteArrayOutputStream, value: Any?) {
        if (value == null) {
            super.writeValue(stream, value)
            return
        }

        val type = when(value) {
            is Messaging.Timestamp -> TIMESTAMP
            is Messaging.Contact -> CONTACT
            is Messaging.Message -> MESSAGE
            is Messaging.Conversation -> CONVERSATION
            else -> 0
        }

        if (type == 0.toByte()) {
            super.writeValue(stream, value)
            return
        }

        stream.write(type.toInt())
        val bytes = (value as GeneratedMessageLite<*, *>).toByteArray()
        writeBytes(stream, bytes)
    }

    override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any {
        if (type < 30) {
            return super.readValueOfType(type, buffer)
        }

        val length = readSize(buffer)
        val serialized = ByteArray(length)
        buffer.get(serialized)
        return when (type) {
            TIMESTAMP -> Messaging.Conversation.parseFrom(serialized)
            CONTACT -> Messaging.Contact.parseFrom(serialized)
            MESSAGE -> Messaging.Message.parseFrom(serialized)
            CONVERSATION -> Messaging.Conversation.parseFrom(serialized)
            else -> throw FormatException("Unknown data type: $type")
        }
    }

    companion object {
        const val TIMESTAMP:Byte = 30
        const val CONTACT:Byte = 31
        const val MESSAGE:Byte = 32
        const val CONVERSATION:Byte = 33
    }
}